### PR TITLE
[wip] First setup custom logging.

### PR DESCRIPTION
**Pull request by 304NotModified**
Code idea, to continue discussion started here: https://winscp.net/forum/viewtopic.php?t=26645

This is a pure refactor, no changes are made in the logging code.

- moved all logging stuff of Logging folder
- created ILogWriterFactory and ILogWriter  interfaces
- created default implementations of LogWriterFactory and ILogWriter, that has the exact same logic as before. (pure refactor)

todo

- [ ] remove filePath logic in Logger (as custom logger is probably not writing to a file)
- [ ] exception and arguments should be passed to the ILogWriter (e.g. NLog could format itself and do some smart lazyiness)
- [ ] Locking should be responsibility of ILogWriter - we don't need it for NLog, as writing is threadsafe

### CODE REVIEW

**Comment by 304NotModified** [comment]
recommend to ignore whitespaces, as .gitattribute isn't set-up (and so VS2017 changes whitespaces)

image

File: dotnet/Logging/FileLogWriter.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Globalization;
+using System.IO;
+using System.Threading;
+
+namespace WinSCP.Internal
+{
+    internal class FileLogWriter : ILogWriter
```
Comment (304NotModified) [inline dotnet/Logging/FileLogWriter.cs]: note: the code in this class is just a move, no functional change.

File: dotnet/LogWriterFactory.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using WinSCP.Internal;
+
+namespace WinSCP
+{
+    class LogWriterFactory
+    {
+        public static ILogWriterFactory CurrentFactory { get; set; } = new FileLogWriterFactory();
```
Comment (304NotModified) [inline dotnet/LogWriterFactory.cs]: note: this is the starting point to inject the ILogWriterFactory

File: dotnet/Logging/Logger.cs [hunk 3]
```diff
@@ -7,14 +7,21 @@
 using System.Collections.Generic;
 using System.ComponentModel;
 using System.Runtime.InteropServices;
+using WinSCP.Internal;
 
 namespace WinSCP
 {
     internal class Logger : IDisposable
```
Comment (304NotModified) [inline dotnet/Logging/Logger.cs]: note: maybe rename of Logger would be better, but for simplicity I tried to make the changes a small as possible

File: dotnet/LogWriterFactory.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using WinSCP.Internal;
+
+namespace WinSCP
+{
+    public static class LogWriterFactory
+    {
+        public static ILogWriterFactory CurrentFactory { get; set; } = new FileLogWriterFactory();
```
Comment (304NotModified) [inline dotnet/LogWriterFactory.cs]: note: this is the starting point to inject the ILogWriterFactory

**Comment by martinprikryl** [comment]
Please:
- Move the `*LogReader` classes back - Those are not related to the logging at all
- Lowercase `Logging` folder
- Use just one empty line, not two, everywhere in FileLogWriter.cs
- Remove `<inheritdoc />`
- Add missing newlines as the end of FileLogWriterFactory.cs, FileLogWriter.cs and ILogWriterFactory.cs
- Change `namespace WinSCP.Internal` to `namespace WinSCP`

**Comment by 304NotModified** [comment]
Will do, but have some questions:

> * Lowercase `Logging` folder

Why is that? The standard naming guidelines for namespaces is PascalCase and so it's advised to keep the namespace and folder in sync


> * Add missing newlines as the end of FileLogWriterFactory.cs, FileLogWriter.cs and ILogWriterFactory.cs

Why do we need that?

**Comment by martinprikryl** [comment]
> Will do, but have some questions:
> 
> > * Lowercase `Logging` folder
> 
> Why is that? The standard naming guidelines for namespaces is PascalCase and so it's advised to keep the namespace and folder in sync

OK, that's possible. But other folders are lower case (while indeed inconsistently PascalCase in the project file). I may fix that eventually. But please make "logging" consistent with other folders for now.

> > * Add missing newlines as the end of FileLogWriterFactory.cs, FileLogWriter.cs and ILogWriterFactory.cs
> 
> Why do we need that?

It's our coding standard.

Thanks.

**Comment by 304NotModified** [comment]
OK clear!

**Comment by martinprikryl** [comment]
Closing, as the author has not responded.

**Comment by 304NotModified** [comment]
Or just busy ;)

But fine, I will create a new PR when I found time.

**Pull request closed without merging**